packages/core/src/render3/animations/css_transition_animator.ts

```diff
@@ -0,0 +1,278 @@
+ * 3. Run a reflow and call requestAnimationFrame()
+ * 4. Then repeat from step 1 and apply the next transition effect
+ *
+ * Once the classes/styles and transition strings have all been applied
+ * then the player code will wait for `transitionend` and will then
+ * only finish the animation player once the longest transition animation
+ * has finished.
+ */
+export class CssTransitionAnimator implements Animator {
+  state: AnimatorState = AnimatorState.Idle;
+
+  private _listeners: (() => any)[] = [];
+  private _effects: Effect[] = [];
+  private _collectedClasses: {[className: string]: boolean}|null;
+  private _collectedStyles: {[key: string]: any}|null;
+  private _captureFn: (event: AnimationEvent) => any;
+  private _startTime = 0;
+  private _maxTime = 0;
+  private _timer: any;
+  private _currentTransitionStr: string = '';
+  private _waitingForFrameFns: (() => any)[] = [];
+
+  constructor(private _element: HTMLElement, collectStyling?: boolean) {
+    this._captureFn = (event: AnimationEvent) => {
+      const totalTime = event.timeStamp - this._startTime;
+      if (event.target === this._element) {
+        event.stopPropagation();
+        if (totalTime >= this._maxTime) {
+          this._onAllEffectsFinished();
+        }
+      }
+    };
+    this._element.addEventListener('transitionend', this._captureFn, {capture: true});
+    this._collectedClasses = collectStyling ? {} : null;
+    this._collectedStyles = collectStyling ? {} : null;
+  }
+
+  onAllEffectsDone(cb: () => any) { this._listeners.push(cb); }
+
+  addEffect(effect: Effect) {
+    const elapsedTimeSoFar = this.state === AnimatorState.Running ? (now() - this._startTime) : 0;
+    const time = elapsedTimeSoFar + effect.timing.duration + (effect.timing.delay || 0);
+    this._maxTime = Math.max(this._maxTime, time);
+    this._effects.push(effect);
+  }
+
+  finishEffect(effect: Effect) {
+    setTransition(this._element, CANCEL_NEXT_TRANSITION_VALUE);
+
+    // the effect has to be first removed and immediately applied right
+    this._applyEffectStyling(effect, true);
+    this._waitForFrame(() => {
+      this._applyEffectStyling(effect, false);
+      this._waitForFrame(() => { this._flushNextEffect(); });
+    });
+  }
+
+  destroyEffect(effect: Effect) {
+    setTransition(this._element, CANCEL_NEXT_TRANSITION_VALUE);
+    this._applyEffectStyling(effect, true);
+    this._waitForFrame(() => this._flushNextEffect());
+  }
+
+  private _applyEffectStyling(effect: Effect, revert: boolean) {
+    effect.classes &&
+        applyClassChanges(this._element, effect.classes, revert, this._collectedClasses);
+    effect.styles &&
+        applyStyleChanges(this._element, effect.styles, null, revert, this._collectedStyles);
+  }
+
+  private _waitForFrame(cb: () => any) {
+    if (this._waitingForFrameFns.length == 0) {
+      applyReflow(this._element, () => {
+        for (let i = 0; i < this._waitingForFrameFns.length; i++) {
+          cb();
```

Comment:
This looks broken? I expect smth like:
```suggestion
          this._waitingForFrameFns[i]();
```

But then `cb` wouldn't be exectuted at all. I am a bit puzzled by this piece of logic.